pkg/templateservicebroker/brokerapi/templateservicebroker/bind.go

```diff
@@ -0,0 +1,105 @@
+)
+
+// copied from vendor/k8s.io/kubernetes/pkg/kubelet/envvars/envvars.go
+func makeEnvVariableName(str string) string {
+	return strings.ToUpper(strings.Replace(str, "-", "_", -1))
+}
+
+func (b *Broker) getServices(namespace, instance_id string) (map[string]interface{}, *brokerapi.Response) {
+	requirement, _ := labels.NewRequirement(templateservicebroker.ServiceInstanceLabel, selection.Equals, []string{instance_id})
+
+	serviceList, err := b.kc.Services(namespace).List(api.ListOptions{LabelSelector: labels.NewSelector().Add(*requirement)})
+	if err != nil {
+		return nil, &brokerapi.Response{http.StatusInternalServerError, nil, err}
+	}
+
+	services := map[string]interface{}{}
+	for _, service := range serviceList.Items {
+		if !api.IsServiceIPSet(&service) || len(service.Spec.Ports) == 0 {
+			continue
+		}
+
+		name := makeEnvVariableName(service.Name) + "_SERVICE_HOST"
+		services[name] = service.Spec.ClusterIP
+
+		name = makeEnvVariableName(service.Name) + "_SERVICE_PORT"
+		services[name] = strconv.Itoa(int(service.Spec.Ports[0].Port))
+
+		for _, port := range service.Spec.Ports {
+			if port.Name != "" {
+				services[name+"_"+makeEnvVariableName(port.Name)] = strconv.Itoa(int(port.Port))
+			}
+		}
+	}
+
+	return services, nil
+}
+
+func (b *Broker) getSecrets(namespace, instance_id string) (map[string]interface{}, *brokerapi.Response) {
+	requirement, _ := labels.NewRequirement(templateservicebroker.ServiceInstanceLabel, selection.Equals, []string{instance_id})
+
+	secretList, err := b.kc.Secrets(namespace).List(api.ListOptions{LabelSelector: labels.NewSelector().Add(*requirement)})
+	if err != nil {
+		return nil, &brokerapi.Response{http.StatusInternalServerError, nil, err}
+	}
+
+	secrets := map[string]interface{}{}
+	for _, secret := range secretList.Items {
+		if secret.Type != api.SecretTypeBasicAuth {
+			continue
+		}
+
+		name := makeEnvVariableName(secret.Name + "_" + api.BasicAuthUsernameKey)
+		secrets[name] = string(secret.Data[api.BasicAuthUsernameKey])
+
+		name = makeEnvVariableName(secret.Name + "_" + api.BasicAuthPasswordKey)
+		secrets[name] = string(secret.Data[api.BasicAuthPasswordKey])
+	}
+
+	return secrets, nil
+}
+
+func (b *Broker) Bind(instance_id, binding_id string, breq *brokerapi.BindRequest) *brokerapi.Response {
+	// TODO: in principle, bind should alter state somewhere
+	um, err := b.oc.UUIDMaps().Get(instance_id)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return &brokerapi.Response{http.StatusBadRequest, nil, err}
+		}
+
+		return &brokerapi.Response{http.StatusInternalServerError, nil, err}
+	}
+
+	namespace := um.Spec.TemplateProvisionRef.Namespace
+
+	services, resp := b.getServices(namespace, instance_id)
+	if resp != nil {
+		return resp
+	}
+
+	secrets, resp := b.getSecrets(namespace, instance_id)
+	if resp != nil {
+		return resp
+	}
+
+	credentials := map[string]interface{}{}
+	// TODO: confirm this API
+	// TODO: we're somewhat overloading 'credentials' here
```

Comment:
i think this is right..I think in the broker api, "credentials" includes service coordinates like a url, so including the service definition is reasonable.